In [31]:
!pip install Bio

### Code for PB1 protein

In [32]:
from Bio import pairwise2
from Bio.pairwise2 import format_alignment
from Bio.Seq import Seq

#pb1 
print("#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018")

# Define sequences
protein_seq2022 = Seq("MDVNPTLLFLKVPAQDAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLETMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTVEVFRSNSLTANESGRLIDFLKDVMDSMDKEEMEITTHFQRKRRVRDNMTKKMVTQRTIGRKKQRLNKRSYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNVLSIAPIMFSNKMARLGKGYMFESKSMKLRTQIPAEMLANIDLKYFNESTRKKIEKIRPLLIDGTVSLSPGMMMGMFNMLSTVLGVSILNLGQKKYTKTTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINRTGTFEFTSFFYRYGFVANFSMELPSFGVSGINESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWEQTRSKAGLLVSDGGPNLYNIRNLHIPEVCLKWDLMDEDYQGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCSLFEKFFPSSSYRRPVGISSMVEAMVSRARIDARIDFESGRIKKEEFAEIMKICSTIEELRRQK*MEQGQDTQWTQSTEHINTQRRENGQQTQKPEHLNSTQLMGHYLRTTNRADMHKQIACWKQWLSLRSPTQGSLKTLVLKRWKSFSKQEWTS*")
protein_seq2021 = Seq("MDVNPTLLFLKVPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLETMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTIEVFRSNGLKANDSGRLIDFLKDVMESMDKEEMEITTHFQRKRRVRDNVTKKMVTQRTIGKKKQRLNKRSYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNVLSIAPIMFSNKMARLGKGYMFESKSMKLRTQIPAEMLANIDLKYFNESTRKKIEKIRPLLIDGTASLSPGMMMGMFNMLSTVLGVSILNLGQKRYTKTTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINRTGTFEFTSFFYRYGFVANFSMELPSFGVSGINESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWEQTRSKAGLLVSDGGPNLYNIRNLHIPEVCLKWELMDEDYQGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCSLFEKFFPSSSYRRPVGISSMVEAMVSRARIDARIDFESGRVKKEEFAEIMKICSTIEELRRQK*")
protein_seq2018 = Seq("MDVNPTLLFLKVPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLEAMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTIEVFRSNGLTASESGRLIDFLRDVMESMDKEEMEITTHFQRKRRVRDNITKKMVTQRTIGKKKQRLNKRSYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNVLSIAPIMFSNKMARLGKGYMFESKSMKLRTQIPAEMLTNIDLKYFNESTRKKIEKIRPLLIDGTASLSPGMMMGMFNMLSTVLGVSILNLGQKRYTKTTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINRTGTFEFTSFFYRYGFVANFSMELPSFGVSGINESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDAQIQTRRSFELKKLWEQTRSKAGLLVSDGGPNLYNIRNLHIPEVCLKWELMDEDYQGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCSLFEKFFPSSSYRRPVGISSMVEAMVSRARIDARIDFESGRIKKEEFAEIMKICSTIEELRRQK*")

# Align Protein2022 with Protein2021
alignments_1_2 = pairwise2.align.globalxx(protein_seq2022, protein_seq2021)
print("Alignment between Protein2022 with Protein2021:")
print(format_alignment(*alignments_1_2[0]))

# Align Protein2022 with Protein2018
alignments_1_3 = pairwise2.align.globalxx(protein_seq2022, protein_seq2018)
print("Alignment between Protein2022 with Protein2018:")
print(format_alignment(*alignments_1_3[0]))

# Align Protein2021 with Protein2018
alignments_2_3 = pairwise2.align.globalxx(protein_seq2021, protein_seq2018)
print("Alignment between Protein2021 with Protein2018:")
print(format_alignment(*alignments_2_3[0]))


#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018
Alignment between Protein2022 with Protein2021:
MDVNPTLLFLKVPAQD-AISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEESHPGIFENSCLETMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTV-EVFRSNS-LT-ANE-SGRLIDFLKDVMD-SMDKEEMEITTHFQRKRRVRDNM-TKKMVTQRTIGRKK-QRLNKRSYLIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNVLSIAPIMFSNKMARLGKGYMFESKSMKLRTQIPAEMLANIDLKYFNESTRKKIEKIRPLLIDGTV-SLSPGMMMGMFNMLSTVLGVSILNLGQKK-YTKTTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRTCKLVGINMSKKKSYINRTGTFEFTSFFYRYGFVANFSMELPSFGVSGINESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWEQTRSKAGLLVSDGGPNLYNIRNLHIPEVCLKWD-LMDEDYQGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQKCCSLFEKFFPSSSYRRPVGISSMVEAMVSRARIDARIDFESGRI-KKEEFAEIMKICSTIEELRRQK*MEQGQDTQWTQSTEHINTQRRENGQQTQKPEHLNSTQLMGHYLRTTNRADMHKQIACWKQWLSLRSPTQGSLKTLVLKRWKSFSKQEWTS*
||||||||||||||

### Code for PB2 Protein

In [36]:
from Bio import pairwise2
from Bio.pairwise2 import format_alignment
from Bio.Seq import Seq

#pb2 
print("#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018")

# Define sequences
protein_seq2022 = Seq("MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMKYPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITKEKKEELQDCKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGEVRNDDVDQSLIIAARNIVRRATVSADPLASLLEMCHSTQIGGIRMVDILRQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKREEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGRDEQSIAEAIIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSRMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTEKLTITYLSSMMWEINGPESVLVNTYQWIIRNWETVKIQWSQDPTMLYNRMEFEPFQSLVPKAARGQYSGFVRTLFQQMRDVLGTFDTVQIIKLLPFAAAPPEQSRMQFSSLTVNVRGSGMRILVRGNSPVFNYNKATKRLIVLGKDAGALTEDPGEGTAGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*")
protein_seq2021 = Seq("MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMRYPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDTNPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITREKKKELQGCKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGVVRNDDVDQSLIIAARNIVRRATVSADPLASLLEMCHSTQIGGVRMVDILKQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKKEEEVLTGNLQTLKIRIHEGYEEFTMVGRRATAILRKATRRLIQLIVSGRDEQSIAEAVIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSKMGVDEYSSTERVVVSIDRFLRVRDQKGNVLLSPEEVSEAQGTEKLTITYSSSMMWEINGPESVLVNTYQWIIRNWETVKIQWSQDPTMLYNKMEFEPFQSLVPKAARGQYSGFVRTLFQQMRDVLGTFDTVQIIKLLPFAAAPPEQSKMQFSSLTVNVRGSGLRILVRGNSPVFNYNRATKRLSILGKDAGALTEDPDEGTSGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDIVLVMKRKRDSSILTDSQTATKRIRMAIN*")
protein_seq2018 = Seq("MERVKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMKYPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITKEKKEELQDCKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGDVRNDDVDQSLIIAARNIVRRAVVSADPLASLLEMCHSTQIGGIRMVDILRQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKREEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGKDEQSIAEAIIVAMVFSQEDCMVKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSKMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTEKLTITYSSSMMWEINGPESVLVNTYQWIIRNWETVKIQWSQDPTMLYNKMEFEPFQSLVPKAARGQYSGFVRTLFQQMRDVLGTFDTVQIIKLLPFAAAPPEQSRMQFSSLTVNVRGSGMRILVRGNSPVFNYNKATKRLTVLGKDAGALTEDPDEGTAGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*")

#protein 1 = 2023 
#protein 2 = 2021 
#protein 3 = 2022

# Align Protein2022 with Protein2021
alignments_1_2 = pairwise2.align.globalxx(protein_seq2022, protein_seq2021)
print("Alignment between Protein2022 with Protein2021:")
print(format_alignment(*alignments_1_2[0]))

# Align Protein2022 with Protein2018
alignments_1_3 = pairwise2.align.globalxx(protein_seq2022, protein_seq2018)
print("Alignment between Protein2022 with Protein2018:")
print(format_alignment(*alignments_1_3[0]))

# Align Protein2021 with Protein2018
alignments_2_3 = pairwise2.align.globalxx(protein_seq2021, protein_seq2018)
print("Alignment between Protein2021 with Protein2018:")
print(format_alignment(*alignments_2_3[0]))


#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018
Alignment between Protein2022 with Protein2021:
MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMK-YPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDI-NPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITK-EKKE-ELQD-CKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGEV-RNDDVDQSLIIAARNIVRRATVSADPLASLLEMCHSTQIGGI-RMVDILR-QNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKR-EEEVLTGNLQTLKIRV-HEGYEEFTMVGRRATAILRKATRRLIQLIVSGRDEQSIAEAI-IVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSR-MGVDEYSSTERVVVSIDRFLRVRDQR-GNVLLSPEEVSET-QGTEKLTITYLSS-MMWEINGPESVLVNTYQWIIRNWETVKIQWSQDPTMLYNR-MEFEPFQSLVPKAARGQYSGFVRTLFQQMRDVLGTFDTVQIIKLLPFAAAPPEQSR-MQFSSLTVNVRGSGM-RILVRGNSPVFNYNK-ATKRL-IVLGKDAGALTEDPG-EGTA-GVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDV-VLVMKRKRDSSILTDSQTATKRIRMAIN*
|||||||||||||||||||||||||||||||||||||||||||||||||||||  |||||||||||||||||||||||||||||||||||||

### Code for NA protein

In [1]:
from Bio import pairwise2
from Bio.pairwise2 import format_alignment
from Bio.Seq import Seq

#NA
print("#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018")

# Define sequences
protein_seq2022 = Seq("MNPNQRIITTGSICMVIGIVSLMLQIGNIISIWVSHSIQTGNQYQPEPCNQSIIAYENNTWINQTYVNINNTNFLAEQAVTSVTLAGNSSLCPISGWAIYSKDNGIRIGSKGDVFVIREPFISCSHLECRTFFLTQGALLNDKHSNGTVKDRSPYRTLMSCPVGEAPSPYNSRFESVAWSASACHDGISWLTIGISGPDNGAVAVLKYNGIITDTIKSWRNNILRTQESECACVNGSCFTVMTDGPSNGQASYKIFKIEKGKVVKSVELNAPNYHYEECSCYPDAGEIMCVCRDNWHGSNRPWVSFNQNLEYQIGYICSGVFGDNPRPNDGTGSCSPMSSNGAYGVKGFSFKYGNGVWIGRTKSTSSRSGFEMIWDPNGWTETDSSFSVKQDIVAITDLSGYSGTFVQHPELTGLDCMRPCFWVELIRGRPKENTVWTSGSSISFCGVNSDTVGWSWPDGAELPFTIDK*")
protein_seq2021 = Seq("MNPNQKIVIVGSISLGLVVFNVLLHAMSIILMVLALGKSENSGICKGTTVREYNETVRIEKVTQWYNTSVVEYVQHWNEGAYTNNTETICDVKGFAPFSKDNGIRVGSRGHIFVIREPFVSCSPIECRTFFLTQGALLNDKHSNGTVKDRSPFRTLMSVEVGQSPNVYQARFEAVAWSATACHDGKKWMTIGVTGPDPKAVAVVHYGGVPTDVVNSWAGDILRTQESSCTCIQGNCYWVMTDGPANRQAQYRIYKANQGKIIDQTDVSFSGGHIEECSCYPNDGKVECVCRDNWTGTNRPVLVISPDLSYRVGYLCAGLPSDTPRGKDAQFVGSCTSPMGNQGYGVKGFGFRQGTDVWVGRTISRTSRSGFEIIRIKNGWTQTSKEQIRRQVVIDNLNWSGYSGSFTLPVELSGRECLVPCFWVEMIRGRPEERTIWTSSSSIVMCGVDHEIADWSWHDGAILPFDIDKM*")
protein_seq2018 = Seq("MNPNQKVICISATGMTLSVVSLLIGIANLGLNIGLHYKMGDAPTVDIPSMNGTNSTTTIINNNTQNNFTNITNIIINKEEERIFLNLTKPLCEVNSWHILSKDNAIRIGEDAHILVTREPYLSCDPQSCRMFALSQGTTLRGRHANGTIHDRSPFRALVSWEMGQAPSPYNVRVECIGWSSTSCHDGISRMSICMSGPNNNASAVVWYGGRPVTEIPSWAGNILRTQESECVCHKGVCPVVMTDGPANNRAATKIIYFKEGKIQKIEELRGNAQHIEECSCYGAVRVIKCVCRDNWKGANRPVITIDPEMMTHTSKYLCSRVLTDTSRPNDPTSGNCDAPIIGGSPDPGVKGFAFLDGENSWLGRTISKDSRSGYEILKVPNAETDTQSGPTSHQIIVNNPNWSGYSGAFIDYWANKECFNPCFYVELIRGRPKESSVLWTSNSIVALCGSKERLGSWSWHDGAEIIYFK*")


#protein 1 = 2023 
#protein 2 = 2021 
#protein 3 = 2022

# Align Protein2022 with Protein2021
alignments_1_2 = pairwise2.align.globalxx(protein_seq2022, protein_seq2021)
print("Alignment between Protein2022 with Protein2021:")
print(format_alignment(*alignments_1_2[0]))

# Align Protein2022 with Protein2018
alignments_1_3 = pairwise2.align.globalxx(protein_seq2022, protein_seq2018)
print("Alignment between Protein2022 with Protein2018:")
print(format_alignment(*alignments_1_3[0]))

# Align Protein2021 with Protein2018
alignments_2_3 = pairwise2.align.globalxx(protein_seq2021, protein_seq2018)
print("Alignment between Protein2021 with Protein2018:")
print(format_alignment(*alignments_2_3[0]))


#protein 1 = 2023 aka 2022 , #protein 2 = 2021 , #protein 3 = 2022 aka 2018
Alignment between Protein2022 with Protein2021:
MNPNQRIITTGS-ICMVI-G-IVSLMLQIG----NIISIWVS--H--SIQT-------GNQYQP--EPCNQSI-IAY-------EN-N-TW--IN----Q---T----YV---NI-----NNTNFLAEQAVTSVTLAGNSSLCPIS----GW-AIY--SKDNGIRI-GSK-GDV--FVIREPFI-SCSHL--ECRTFFLTQGALLNDKHSNGTVKDRSPY-RTLMSCPVGEAP---SPYNS----RFES-VAWSAS-ACHDGIS--WL-TIGIS--GPDNG--AVAVLK--YNGII---TDTIK---SWRNN---ILRTQESE-CA-CVN--GS-CFT--VMTDGPS-NG-QAS-YKIFK-IE-K---GKV------VKSVELNAPNY----HY-EECSCYP-DAG--EIMCVCRDNWH-GS-NRPWV---SFNQN--LE-YQI--GYI-CSGVF-G---DN-PRPNDGT------GSC-SPMSS-N-GAYGVKGFSFKYGN---G--VWI-GRTK-S-TSSRSGFEMIWDP----NGWTE-TDSSFSVK-QDI---VA-ITD-L--SGYSGT-FVQH--P-ELT-GLDCMR----PCFWVEL-IRGRPKEN--TV-WTSGSSISF---CGVNSDTVG-----WSWP-DGAE-LPFT-IDK-*
|||||        |  || | | |  |  |    |     |   |  ||         |       |  | |  |         |  | |   |     |   |    ||   |      |||    |      |         |     |  |    |||||||  ||  |    |||||||  |||    ||||||||||||||||||||

/srv/conda/lib/python3.11/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.
  warnings.warn(


#### Code for HA protein


In [2]:
from Bio import pairwise2
from Bio.pairwise2 import format_alignment
from Bio.Seq import Seq

print("HA protein analysis #protein 1 = 2023 aka 2022, #protein 2 = 2021, #protein 3 = 2022 aka 2018")

# Define sequences
protein_seq2022 = Seq("MENIVLLLAIISLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKAHNGKLCDLNGVKPLILKDCSVAGWLLGNPMCDEFIRVPEWSYIVERANPANDLCYPGSLNDYEELKHLLSRINHFEKILIIPKSSWPNHETSLGVSAACPYQGAPSFFRNVVWLIKKNDAYPTIKISYNNTNREDLLILWGIHHSNNAEEQTNLYKNPTTYISVGTSTLNQRLVPKIATRSQVNGQRGRMDFFWTILKPDDAIHFESNGNFIAPEYAYKIVKKGDSTIMKSGVEYGHCNTKCQTPVGAINSSMPFHNIHPLTIGECPKYVKSNKLVLATGLRNSPLRERRRKRGLFGAIAGFIEGGWQGMVDGWYGYHHSNEQGSGYAADKESTQKAIDGVTNKVNSIIDKMNTQFEAVGREFNNLERRIENLNKKMEDGFLDVWTYNAELLVLMENERTLDFHDSNVKNLYDKVRLQLRDNAKELGNGCFEFYHKCDDECMESVRNGTYDYPQYSEEARLKREEISGVKLESIGTYQILSIYSTAASSLALAIMMAGLSLWMCSNGSLQCRICI")
protein_seq2021 = Seq("MERTVLLLAIVSLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKAHNGKLCDLNGVKPLVLNDCSVAGWLLGNPMCDEFIRVPEWSYIVERTNPANDLCYPGSLNDYEELKHLLSRIKHFEKILIIPKSSWPNHETSLGVSAACPYQGTPSFFRNVVWLIKKNDAYPTIKINYNNTNREDLLILWGIHHSNNEEEQTNLYKNPTTYISVGTSTLNQRMVPKIATRSQVNGQRGRMDFFWTILKPNDAIHFESNGNFIAPEYAYKIVKKGDSTIMKSEVEYGHCNTKCQTPVGAINSSLPFHNIHPLTIGECPKYVKSNKLVLATGLRNSPLREKRRKRGLFGAIAGFIEGGWQGMVDGWYGYHHSNEQGSGYAADKESTQKAIDGVTNKVNSIINKMNTQFEAVGREFNNLERRIENLNKKMEDGFLDVWTYNAELLVLMENERTLDFHDSNVKNLYDKVRLQLRDNAKELGNGCFEFYHKCDNECMESVRNGTYYYPQYSEEARLKREEISGVKLESIGTYQILSIYSTVASSLALAIMVAGLSLWMCSNGSLQCRICI")
protein_seq2018 = Seq("MENIVLLLAIVSLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKTHNGKLCDLNGVKPLILKDCSVAGWLLGNPMCDEFIRVPEWSYIVERDNPANDLCYPGSLNDYEELKHLLSRINHFEKILIIPKSSWPNHETSLGVSAACPYQGTPSFFRNVVWLIKKNDAYPTIKISYNNTNREDLLILWGIHHSNNAEEQTNLYKNPTTYISVGTSTLNQRLVPKIATRSQVNGQRGRMDFFWTILKPNDAIHFESNGNFIAPEYAYKIVKKGDSTIMKSGVEYGHCNTKCQTPVGAINSSMPFHNIHPLTIGECPKYVKSNKLVLATGLRNSPLREKRRKRGLFGAIAGFIEGGWQGMVDGWYGYHHSNEQGSGYAADKESTQKAIDGVTNKVNSIIDKMNTQFEAVGREFNNLERRIENLNKKMEDGFLDVWTYNAELLVLMENERTLDFHDSNVKNLYDKVRLQLRDNAKELGNGCFEFYHKCDNECMESVRNGTYDYPQYSEEARLKREEISGVKLESIGTYQILSIYSTVASSLALAIMVAGLSLWMCSNGSLQCRICI")

# Align Protein2022 with Protein2021
alignments_1_2 = pairwise2.align.globalxx(protein_seq2022, protein_seq2021)
print("Alignment between Protein2022 with Protein2021:")
print(format_alignment(*alignments_1_2[0]))

# Align Protein2022 with Protein2018
alignments_1_3 = pairwise2.align.globalxx(protein_seq2022, protein_seq2018)
print("Alignment between Protein2022 with Protein2018:")
print(format_alignment(*alignments_1_3[0]))

# Align Protein2021 with Protein2018
alignments_2_3 = pairwise2.align.globalxx(protein_seq2021, protein_seq2018)
print("Alignment between Protein2021 with Protein2018:")
print(format_alignment(*alignments_2_3[0]))

HA protein analysis #protein 1 = 2023 aka 2022, #protein 2 = 2021, #protein 3 = 2022 aka 2018
Alignment between Protein2022 with Protein2021:
MENI--VLLLAII-SLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKAHNGKLCDLNGVKPLI-LK-DCSVAGWLLGNPMCDEFIRVPEWSYIVERA-NPANDLCYPGSLNDYEELKHLLSRIN-HFEKILIIPKSSWPNHETSLGVSAACPYQGA-PSFFRNVVWLIKKNDAYPTIKIS-YNNTNREDLLILWGIHHSNNAEE-QTNLYKNPTTYISVGTSTLNQRL-VPKIATRSQVNGQRGRMDFFWTILKPD-DAIHFESNGNFIAPEYAYKIVKKGDSTIMKSG-VEYGHCNTKCQTPVGAINSSM-PFHNIHPLTIGECPKYVKSNKLVLATGLRNSPLRER-RRKRGLFGAIAGFIEGGWQGMVDGWYGYHHSNEQGSGYAADKESTQKAIDGVTNKVNSIID-KMNTQFEAVGREFNNLERRIENLNKKMEDGFLDVWTYNAELLVLMENERTLDFHDSNVKNLYDKVRLQLRDNAKELGNGCFEFYHKCDD-ECMESVRNGTYDY-PQYSEEARLKREEISGVKLESIGTYQILSIYSTA-ASSLALAIMM-AGLSLWMCSNGSLQCRICI
||    ||||| | |||||||||||||||||||||||||||||||||||||||||||||||||||||||  |  |||||||||||||||||||||||||||||  |||||||||||||||||||||||||  ||||||||||||||||||||||||||||||  ||||||||||||||||||||||  |||||||||||||||||||| || ||||||||||||||||||||||  ||||||||||||||||||||||||||  ||||||

## Analysis From Pairwise Alignment Scores and Patterns

#### PB1 Analysis Insights from Pairwise Alignments

The pairwise alignment analysis of the PB1 protein from influenza viruses over three different years (2022, 2021, 2018) reveals a high degree of conservation, with alignment scores ranging from 744 to 749. This conservation is consistent with PB1's critical role in viral RNA synthesis. The alignments show that most differences between the sequences are single amino acid substitutions, with a notable single amino acid insertion/deletion near the beginning of the sequence (position 15) in the 2022 vs 2021 and 2022 vs 2018 comparisons. 

The C-terminal region displays some variability, with additional amino acids present in the 2022 sequence compared to both 2021 and 2018.
The high similarity between sequences suggests a relatively slow evolutionary rate for PB1, which is expected given its essential function in viral replication. This conservation implies strong functional constraints on the protein, with perfectly conserved regions likely representing critical functional domains, such as the four conserved motifs essential for RNA-dependent RNA polymerase activity. Areas with more frequent substitutions or indels may represent regions under less functional constraint or possibly subject to immune pressure. The additional amino acids in the 2022 sequence at the C-terminus could indicate a recent evolutionary change, possibly involved in protein-protein interactions or regulatory functions.


The end part of the protein (the C-terminus) shows some differences too. The 2022 version has some extra amino acids that aren't in the 2021 or 2018 versions. This could be a recent change in how the virus evolves. We can see this in the alignment:
"STIEELRRQKMEQGQDTQWTQSTEHINTQRRENGQQTQKPEHLNSTQLMGHYLRTTNRADMHKQIACWKQWLSLRSPTQGSLKTLVLKRWKSFSKQEWTS"
"ST--I-----E--------E-----------L----R----------------R---Q---------------K-----*"
The top line shows the longer 2022 sequence, while the bottom line shows the shorter 2021 sequence.


While PB1 is not a primary target for neutralizing antibodies, the observed changes contribute to overall viral evolution and potentially to antigenic drift in influenza viruses. The high conservation of PB1 suggests it could be a potential target for universal influenza vaccines, as it changes less rapidly than surface proteins like hemagglutinin and neuraminidase. This analysis underscores the importance of PB1 in influenza virus biology and its potential as a target for both vaccine development and antiviral therapies. Further investigation into the functional significance of the C-terminal extension in the 2022 (2023) sequence may provide additional insights into PB1's role in viral replication and adaptation.


#### PB2 Analysis Insights from Pairwise Alignments
The pairwise alignment analysis of the PB2 protein from influenza viruses over three different years (2022, 2021, and 2018) shows a high degree of conservation, with alignment scores ranging from 736 to 750. 
This conservation is consistent with PB2's role in viral RNA synthesis. The alignments show that most differences between the sequences are single amino acid substitutions, with a few small insertions or deletions (indels) scattered throughout the sequence.

The highest similarity is observed between the 2022 and 2018 sequences, with an alignment score of 750, suggesting that the 2022 strain may have retained or reverted to some features of the older 2018 strain. The 2021 strain shows slightly more divergence, with lower alignment scores when compared to both 2022 and 2018 (737 and 736). This pattern could indicate a fluctuation in evolutionary pressures or adaptation strategies over the years.

Throughout the sequences, there are several regions of perfect conservation, probably representing important domains for PB2's role in the viral polymerase complex. The presence of small gaps and amino acid substitutions in other areas suggests regions that may be more tolerant to changes, possibly involved in host adaptation or immune evasion.

Some observations include small indels near the N-terminus and scattered throughout the middle of the protein. These changes could potentially affect protein folding and change the protein's function. The C-terminal region shows a high degree of conservation across all three years, indicating its likely importance in maintaining PB2's core functions.

This analysis underscores the evolutionary constraints on the PB2 protein, reflecting its essential role in viral replication. The observed changes, while relatively few, may contribute to the virus's ability to adapt to different host environments or evade host immune responses. When looking at the other 3 proteins we can see what the differences may be. 

#### NA Analysis 

The pairwise alignment analysis of the NA (neuraminidase) protein from influenza viruses over three different years (2022, 2021, and 2018) reveals significant variability and evolutionary changes, especially compared to the other three. 

Overall Conservation and Variability:
The alignment scores indicate substantial differences between the sequences:

2022 vs 2021: Score = 278
2022 vs 2018: Score = 249
2021 vs 2018: Score = 247

These relatively low scores suggest that NA undergoes considerable changes over time, which is consistent with its role as a surface protein subject to immune pressure.


The alignments show numerous gaps throughout the sequences, indicating frequent insertions and deletions. This suggests that the NA protein can tolerate structural changes in certain regions. There are also many positions where amino acids differ between the years, showing that there are ongoing mutations and adaptations. Despite the overall variability, some segments show higher conservation. These likely correspond to functionally critical areas, such as the enzyme active site or key structural elements. The sequence "MNPNQ" at the N-terminus is conserved across all three years.
A segment "CHDG" appears conserved in all three sequences.
The sequence "FVSHKEIESVNNAVVMPAHGPAKSMEYDAVATTHSWIPKRNRSILNTSQRGILEDEQMYQK" towards the C-terminus shows high conservation between 2022 and 2021.

Both 2022 and 2021 sequences show lower similarity to the 2018 sequence (scores 249 and 247 respectively), suggesting significant changes over this time period.
The similar scores of 2022 vs 2018 (249) and 2021 vs 2018 (247) imply that the rate of change from 2018 to 2021 and 2022 was relatively consistent.

The high variability observed is characteristic of antigenic drift, allowing the virus to evade host immune responses. Conserved regions likely play crucial roles in maintaining NA's enzymatic activity, which is essential for viral release from host cells. The high presence of indels suggests that NA has some structural flexibility, potentially allowing it to adapt to different host environments or escape antibody recognition. 

Highly variable regions, especially those on the protein surface, may represent epitopes targeted by the immune system.

In conclusion, this analysis of NA sequences from 2018 to 2022 reveals a protein that is rapidly evolving, with a mix of conserved functional regions and highly variable segments. This pattern of change makes NA a critical component in understanding influenza virus evolution.

#### HA Analysis 


The pairwise alignment analysis of the HA (hemagglutinin) protein from influenza viruses over three different years (2022, 2021, and 2018) reveals patterns of both conservation and variation. 

The alignment scores are:

2022 vs 2021: Score = 547
2022 vs 2018: Score = 558
2021 vs 2018: Score = 553


These high scores suggest that HA maintains a considerable degree of conservation over the observed time period. Large portions of the sequence are identical across all three years, indicating functionally critical regions. There are several positions where amino acids differ between years, showing ongoing mutations and adaptations. There are very few gaps in the alignment, suggesting that the overall length and structure of HA are well-conserved.

When looking at some examples, we can identify conserved and variable regions:

Conserved C-terminal region:
"AGLSLWMCSNGSLQCRICI" is identical in all three years, likely representing a conserved structural or functional domain.

Variable N-terminal region:
"MENI--VLLLAII-SLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKAHNGKLCDLNGVKPLI" (2022)
"ME--RTVLLLA-IVSLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQDILEKAHNGKLCDLNGVKPL" (2021)
We can see that the virus changed the first few amino acids of HA between 2021 and 2022.

Variable middle region:
"NPANDLCYPGSLNDYEELKHLLSRIN-HFEKILIIPKSSWPNHETSLGVSAACPYQGA-PSFFRNVVWLIKKNDAYPTIKIS" (2022)
"TNPANDLCYPGSLNDYEELKHLLSRI-KHFEKILIIPKSSWPNHETSLGVSAACPYQG-TPSFFRNVVWLIKKNDAYPTIKI" (2021)

Here, we can see a few single amino acid changes, like the 'N' to 'K' switch near the beginning of this segment.

The fact that HA shows changes makes sense because it's on the outside of the virus and is a main target for our immune system. The virus needs to keep changing HA to avoid being recognized by antibodies from previous infections or vaccinations.

This analysis of HA sequences from 2018 to 2022 reveals a protein that undergoes significant changes over time while maintaining its core structure and function. The high conservation in large parts of the sequence shows the critical functional roles of HA in viral entry and fusion. The changes observed can be adaptations that enhanced viral fitness or immune evasion.